Remotion/Data/DomainObjects/Mapping/TypeDefinition.cs

```diff
@@ -0,0 +1,248 @@
+      _cachedRelationEndPointDefinitions = new Lazy<RelationEndPointDefinitionCollection>(
+           () => RelationEndPointDefinitionCollection.CreateForAllRelationEndPoints(this, true),
+           LazyThreadSafetyMode.ExecutionAndPublication);
+      _cachedPropertyDefinitions =
+          new Lazy<PropertyDefinitionCollection>(
+              () => new PropertyDefinitionCollection(PropertyDefinitionCollection.CreateForAllProperties(this, true), true),
+              LazyThreadSafetyMode.ExecutionAndPublication);
+    }
+
+    public IStorageEntityDefinition StorageEntityDefinition
+    {
+      get
+      {
+        if (_storageEntityDefinition == null)
+          throw new InvalidOperationException($"StorageEntityDefinition has not been set for type '{Type.GetFullNameSafe()}'.");
+        return _storageEntityDefinition;
+      }
+    }
+
+    public bool HasStorageEntityDefinitionBeenSet => _storageEntityDefinition != null;
+
+    public PropertyDefinitionCollection MyPropertyDefinitions
+    {
+      get
+      {
+        if (_propertyDefinitions == null)
+          throw new InvalidOperationException($"No property definitions have been set for type '{Type.GetFullNameSafe()}'.");
+
+        return _propertyDefinitions;
+      }
+    }
+
+    public RelationEndPointDefinitionCollection MyRelationEndPointDefinitions
+    {
+      get
+      {
+        if (_relationEndPoints == null)
+          throw new InvalidOperationException($"No relation end point definitions have been set for type '{Type.GetFullNameSafe()}'.");
+
+        return _relationEndPoints;
+      }
+    }
+
+    public virtual bool IsTypeResolved => true;
+
+    public abstract bool IsPartOfInheritanceHierarchy { get; }
+
+    public abstract TypeDefinition GetInheritanceRoot ();
+
+    public PropertyDefinition? this [string propertyName]
+    {
+      get
+      {
+        ArgumentUtility.CheckNotNullOrEmpty(nameof(propertyName), propertyName);
+        return MyPropertyDefinitions[propertyName];
+      }
+    }
+
+    /// <summary>
+    /// Returns all types that are part of this type's hierarchy, including itself, using depth-first.
+    /// </summary>
+    public abstract TypeDefinition[] GetTypeHierarchy ();
+
+    protected abstract void CheckPropertyDefinitions (IEnumerable<PropertyDefinition> propertyDefinitions);
+
+    protected abstract void CheckRelationEndPointDefinitions (IEnumerable<IRelationEndPointDefinition> relationEndPoints);
+
+    public void SetStorageEntity (IStorageEntityDefinition storageEntityDefinition)
+    {
+      ArgumentUtility.CheckNotNull(nameof(storageEntityDefinition), storageEntityDefinition);
+
+      if (IsReadOnly)
+        throw new NotSupportedException($"Type '{Type.GetFullNameSafe()}' is read-only.");
+
+      _storageEntityDefinition = storageEntityDefinition;
+    }
+
+    public PropertyDefinitionCollection GetPropertyDefinitions ()
+    {
+      return _cachedPropertyDefinitions.Value;
```

Comment:
I think we need to guard this (and relations) with IsReadOnly and either throw or return a new instance when we're not readonly. otherwise, we could accidently cache incomplete data. It's not there in the original code either, but reading it now, it feels icky

Reply:
This was done in an upcoming commit together with the change on how read-only behaves.